File: android/app/src/main/java/com/neotree/app/net/Schedule.kt
package com.neotree.app.net

import org.json.JSONArray
import org.json.JSONObject
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

/**
 * The tree's schedule (firmware engine/include/neotree/schedule.hpp): lights
 * on/off timers - like a plug-in Christmas light timer, by day of the week -
 * and events at set moments. Times are in the tree's local time.
 */
data class TreeSchedule(val timers: List<TimerInfo>, val events: List<EventInfo>) {
    companion object {
        const val MAX_TIMERS = 8
        const val MAX_EVENTS = 12

        fun parse(json: JSONObject): TreeSchedule {
            val t = json.optJSONArray("timers") ?: JSONArray()
            val e = json.optJSONArray("events") ?: JSONArray()
            return TreeSchedule(
                timers = (0 until t.length()).map { i ->
                    val o = t.getJSONObject(i)
                    TimerInfo(o.optInt("on", 1) == 1, o.optInt("d", EVERY_DAY), o.optInt("a"), o.optInt("b"))
                },
                events = (0 until e.length()).map { i ->
                    val o = e.getJSONObject(i)
                    EventInfo(
                        enabled = o.optInt("on", 1) == 1,
                        name = o.optString("n"),
                        repeat = Repeat.entries.getOrElse(o.optInt("r")) { Repeat.ONCE },
                        year = o.optInt("y", 2026),
                        month = o.optInt("m", 12),
                        day = o.optInt("day", 31),
                        days = o.optInt("d", EVERY_DAY),
                        timeSec = o.optInt("t"),
                        action = EventAction.entries.getOrElse(o.optInt("w")) { EventAction.SHOW },
                        target = o.optString("x"),
                        durationSec = o.optLong("len"),
                    )
                },
            )
        }
    }
}

/** Days of the week as a mask: bit 0 = Sunday ... bit 6 = Saturday. */
const val EVERY_DAY = 0x7F

fun dayBit(d: DayOfWeek): Int = 1 shl (d.value % 7)

/** Lights on at [onSec], off at [offSec] (seconds into the day), starting on [days]; an off at or before the on is the next day. */
data class TimerInfo(val enabled: Boolean = true, val days: Int = EVERY_DAY, val onSec: Int = 17 * 3600, val offSec: Int = 23 * 3600)

enum class Repeat(val label: String) { ONCE("Once"), YEARLY("Every year"), WEEKLY("Every week") }

enum class EventAction(val label: String) { PRESET("Scene"), SHOW("Show"), MODE("Mode") }

data class EventInfo(
    val enabled: Boolean = true,
    val name: String = "",
    val repeat: Repeat = Repeat.YEARLY,
    val year: Int = 2026,
    val month: Int = 12,
    val day: Int = 31,
    val days: Int = EVERY_DAY,
    val timeSec: Int = 23 * 3600 + 59 * 60,
    val action: EventAction = EventAction.SHOW,
    /** A scene or show name, or a mode id. */
    val target: String = "",
    /** Then back to what was playing before; 0 = it stays. */
    val durationSec: Long = 600,
)

/** The running event, from the scene JSON's "event". */
data class RunningEvent(val name: String, val index: Int, val leftSec: Long)

/** What the timer does next: [on] at [at] (local time). */
data class TimerChange(val on: Boolean, val at: LocalDateTime)

object TimerMath {
    /** Whether the timers say on at [t]: true / false, null if none is enabled (the firmware's timer_level). */
    fun level(timers: List<TimerInfo>, t: LocalDateTime): Boolean? {
        val active = timers.filter { it.enabled && it.days and EVERY_DAY != 0 }
        if (active.isEmpty()) return null
        return active.any { r -> windows(r, t.toLocalDate().minusDays(1), t.toLocalDate()).any { (s, e) -> !t.isBefore(s) && t.isBefore(e) } }
    }

    /** The timer's next change after [now] within a week, or null. */
    fun next(timers: List<TimerInfo>, now: LocalDateTime): TimerChange? {
        val current = level(timers, now) ?: return null
        val edges = timers.filter { it.enabled }
            .flatMap { r -> windows(r, now.toLocalDate().minusDays(1), now.toLocalDate().plusDays(8)).flatMap { listOf(it.first, it.second) } }
            .filter { it.isAfter(now) }
            .sorted()
        for (edge in edges) {
            val after = level(timers, edge) ?: return null
            if (after != current) return TimerChange(after, edge)
        }
        return null
    }

    private fun windows(r: TimerInfo, from: LocalDate, to: LocalDate): List<Pair<LocalDateTime, LocalDateTime>> {
        val out = mutableListOf<Pair<LocalDateTime, LocalDateTime>>()
        var d = from
        while (!d.isAfter(to)) {
            if (r.days and dayBit(d.dayOfWeek) != 0) {
                val start = d.atStartOfDay().plusSeconds(r.onSec.toLong())
                val end = d.atStartOfDay().plusSeconds(r.offSec.toLong() + if (r.offSec <= r.onSec) 86400 else 0)
                out += start to end
            }
            d = d.plusDays(1)
        }
        return out
    }
}

/** "5:00 PM", or with seconds when it has them ("11:59:30 PM"). */
fun formatTimeOfDay(sec: Int): String {
    val t = LocalTime.ofSecondOfDay((sec % 86400).toLong())
    val h = if (t.hour % 12 == 0) 12 else t.hour % 12
    val ampm = if (t.hour < 12) "AM" else "PM"
    return if (t.second != 0) "%d:%02d:%02d %s".format(h, t.minute, t.second, ampm) else "%d:%02d %s".format(h, t.minute, ampm)
}

private val DAY_LETTERS = listOf("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat")

/** "Every day", "Weekdays", "Weekends", or "Mon, Wed, Fri". */
fun formatDays(days: Int): String = when (days and EVERY_DAY) {
    EVERY_DAY -> "Every day"
    0b0111110 -> "Weekdays"
    0b1000001 -> "Weekends"
    0 -> "No days"
    else -> (0..6).filter { days shr it and 1 == 1 }.joinToString(", ") { DAY_LETTERS[it] }
}

private val MONTHS = listOf("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")

/** "Every year on Dec 31 at 11:59 PM", "Once on Dec 31 2026 at ...", "Fri, Sat at ...". */
fun formatWhen(e: EventInfo): String {
    val date = MONTHS.getOrElse(e.month - 1) { "?" } + " " + e.day
    val at = formatTimeOfDay(e.timeSec)
    return when (e.repeat) {
        Repeat.ONCE -> "$date ${e.year} at $at"
        Repeat.YEARLY -> "Every $date at $at"
        Repeat.WEEKLY -> "${formatDays(e.days)} at $at"
    }
}

File: engine/src/schedule.cpp
#include "neotree/schedule.hpp"

#include "json.hpp"
#include "neotree/civil_time.hpp"

namespace neotree {

namespace {

using detail::Json;

int64_t floor_div(int64_t a, int64_t b) { return a / b - ((a % b != 0) && ((a < 0) != (b < 0))); }
int weekday_of(int64_t day) { return static_cast<int>(((day % 7) + 11) % 7); }   // 1970-01-01: Thursday

bool terminated(char *s, size_t size)
{
    for (size_t i = 0; i < size; i++)
    {
        if (s[i] == '\0')
        {
            return true;
        }
        if (static_cast<unsigned char>(s[i]) < 0x20)
        {
            return false;
        }
    }
    s[size - 1] = '\0';
    return true;
}

}  // namespace

int timer_level(const Schedule &schedule, int64_t local_s)
{
    const int64_t today = floor_div(local_s, 86400);
    bool any = false;
    for (uint8_t i = 0; i < schedule.timer_count; i++)
    {
        const TimerRule &r = schedule.timers[i];
        if (!r.enabled || (r.days & every_day) == 0)
        {
            continue;
        }
        any = true;
        // A window starting yesterday may still be open (overnight).
        for (int64_t day = today - 1; day <= today; day++)
        {
            if ((r.days >> weekday_of(day) & 1) == 0)
            {
                continue;
            }
            const int64_t start = day * 86400 + r.on_s;
            const int64_t end = day * 86400 + r.off_s + (r.off_s <= r.on_s ? 86400 : 0);
            if (local_s >= start && local_s < end)
            {
                return 1;
            }
        }
    }
    return any ? 0 : -1;
}

bool event_on_day(const ScheduledEvent &e, int64_t day)
{
    switch (e.repeat)
    {
    case Repeat::weekly:
        return (e.days >> weekday_of(day) & 1) != 0;
    case Repeat::yearly:
    case Repeat::once:
    {
        int32_t y;
        uint32_t m, d;
        civil_from_days(day, y, m, d);
        return m == e.month && d == e.day && (e.repeat == Repeat::yearly || y == e.year);
    }
    }
    return false;
}

bool valid_timer(const TimerRule &r)
{
    return r.on_s < 86400 && r.off_s < 86400 && (r.days & ~every_day) == 0;
}

bool valid_event(ScheduledEvent &e)
{
    return terminated(e.name, sizeof(e.name)) && terminated(e.target, sizeof(e.target)) &&
           e.repeat <= Repeat::weekly && e.action <= EventAction::mode && e.time_s < 86400 && e.month >= 1 &&
           e.month <= 12 && e.day >= 1 && e.day <= 31 && (e.days & ~every_day) == 0 && e.target[0] != '\0';
}

size_t describe_schedule(const Schedule &s, uint32_t revision, char *out, size_t cap)
{
    if (cap == 0)
    {
        return 0;
    }
    Json j{out, cap};
    j.raw("{\"rev\":%lu,\"timers\":[", static_cast<unsigned long>(revision));
    for (uint8_t i = 0; i < s.timer_count; i++)
    {
        const TimerRule &r = s.timers[i];
        j.raw("%s{\"on\":%d,\"d\":%u,\"a\":%lu,\"b\":%lu}", i ? "," : "", r.enabled ? 1 : 0,
              static_cast<unsigned>(r.days), static_cast<unsigned long>(r.on_s), static_cast<unsigned long>(r.off_s));
    }
    j.raw("],\"events\":[");
    for (uint8_t i = 0; i < s.event_count; i++)
    {
        const ScheduledEvent &e = s.events[i];
        j.raw("%s{\"on\":%d,\"n\":", i ? "," : "", e.enabled ? 1 : 0);
        j.str(e.name);
        j.raw(",\"r\":%u,\"y\":%d,\"m\":%u,\"day\":%u,\"d\":%u,\"t\":%lu,\"w\":%u,\"x\":",
              static_cast<unsigned>(e.repeat), static_cast<int>(e.year), static_cast<unsigned>(e.month),
              static_cast<unsigned>(e.day), static_cast<unsigned>(e.days), static_cast<unsigned long>(e.time_s),
              static_cast<unsigned>(e.action));
        j.str(e.target);
        j.raw(",\"len\":%lu}", static_cast<unsigned long>(e.duration_s));
    }
    j.raw("]}");
    return j.len;
}

}  // namespace neotree
